File: NyaLauncher.Core/Config/LauncherConfig.cs
using System;
using System.IO;
using System.Linq;
using NyaLauncher.Core.Tools;

namespace NyaLauncher.Core.Config;

/// <summary>
/// 启动器配置的统一入口。底层由 <see cref="ConfigFileManager"/> 负责 JSON 读写。
/// 配置文件为存储目录下的 config.json，默认目录与 workspace.json 保持一致
/// （%USERPROFILE%\NyaLauncher），可通过 <see cref="SetStorageDirectory"/> 同步工作区存储目录。
/// </summary>
public static class LauncherConfig
{
    /// <summary>默认存储目录：%USERPROFILE%\NyaLauncher（与 workspace.json 默认目录一致，便于用户直接找到并编辑）。</summary>
    public static string DefaultStorageDirectory { get; } = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
        "NyaLauncher");

    /// <summary>
    /// 旧版默认存储目录：%LOCALAPPDATA%\NyaLauncher。
    /// 仅用于从旧版本一次性迁移配置到 <see cref="DefaultStorageDirectory"/>。
    /// </summary>
    public static string LegacyDefaultStorageDirectory { get; } = Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
        "NyaLauncher");

    private static readonly object SyncRoot = new();
    private static string _storageDirectory = DefaultStorageDirectory;
    private static ConfigFileManager? _store;

    /// <summary>config.json 所在目录；默认与 workspace.json 同目录。</summary>
    public static string StorageDirectory
    {
        get
        {
            lock (SyncRoot)
                return _storageDirectory;
        }
        private set
        {
            lock (SyncRoot)
                _storageDirectory = value;
        }
    }

    /// <summary>配置文件路径：存储目录下的 config.json。</summary>
    public static string FilePath
    {
        get
        {
            lock (SyncRoot)
                return GetFilePath();
        }
    }

    /// <summary>
    /// 切换 config.json 的读取目录。文件迁移与冲突处理由前端的配置存储协调流程完成；
    /// 本方法只重置底层存储，使后续读取立即应用新目录中的配置。
    /// </summary>
    public static void SetStorageDirectory(string storageDirectory)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(storageDirectory);

        var normalized = Path.TrimEndingDirectorySeparator(Path.GetFullPath(storageDirectory));
        lock (SyncRoot)
        {
            if (PathUtil.PathsEqual(normalized, _storageDirectory))
                return;

            StorageDirectory = normalized;
            _store = null; // 下次访问时用新路径重新加载
        }
    }

    /// <summary>游戏根目录（.minecraft 或自定义目录）；未配置时返回 null。</summary>
    public static string? GameDirectory
    {
        get => WithStore(store =>
        {
            var value = store.MinecraftPathGet();
            return string.IsNullOrWhiteSpace(value) ? null : value;
        });
    }

    /// <summary>首选 Java 可执行文件（java.exe）路径；未配置时返回 null。</summary>
    public static string? JavaExecutable
    {
        get => WithStore(store =>
        {
            var item = store.JavaPathGet().FirstOrDefault();
            return string.IsNullOrWhiteSpace(item.JavaPath) ? null : item.JavaPath;
        });
    }

    /// <summary>首选 Java 的版本号（如 21）；未配置时返回 null。</summary>
    public static string? JavaVersion
    {
        get => WithStore(store =>
        {
            var item = store.JavaPathGet().FirstOrDefault();
            return string.IsNullOrWhiteSpace(item.JavaVersion) ? null : item.JavaVersion;
        });
    }

    /// <summary>保存游戏目录。</summary>
    public static bool SaveGameDirectory(string path) =>
        !string.IsNullOrWhiteSpace(path) &&
        WithStore(store => store.MinecraftPathSet(path.Trim()));

    /// <summary>
    /// 保存首选 Java（java.exe 路径 + 版本）。采用「先清空再写入」策略，
    /// 保证 config.json 中的 javaPath 始终只有一条首选配置。
    /// </summary>
    public static bool SaveJava(string javaPath, string javaVersion)
    {
        if (string.IsNullOrWhiteSpace(javaPath))
            return false;

        return WithStore(store => store.JavaPathSet(
            javaPath.Trim(),
            string.IsNullOrWhiteSpace(javaVersion) ? "unknown" : javaVersion.Trim()));
    }

    /// <summary>删除已保存的 Java 配置（恢复自动检测）。</summary>
    public static void ClearJava() =>
        WithStore(store => store.ConfigItemDelete("javaPath"));

    /// <summary>已保存的全部 Java 路径（列表首位为默认 Java）。</summary>
    public static IReadOnlyList<ConfigFileManager.JavaPathItem> GetJavaPaths() =>
        WithStore(store => store.JavaPathGet());

    /// <summary>添加一条 Java 路径；路径已存在时更新其版本。返回是否成功。</summary>
    public static bool AddJava(string javaPath, string javaVersion)
    {
        if (string.IsNullOrWhiteSpace(javaPath))
            return false;
        return WithStore(store => store.JavaPathAdd(
            javaPath.Trim(),
            string.IsNullOrWhiteSpace(javaVersion) ? "unknown" : javaVersion.Trim()));
    }

    /// <summary>移除一条 Java 路径。返回是否移除成功。</summary>
    public static bool RemoveJava(string javaPath) =>
        !string.IsNullOrWhiteSpace(javaPath) &&
        WithStore(store => store.JavaPathRemove(javaPath.Trim()));

    /// <summary>把指定路径设为默认 Java（列表首位）。返回是否成功。</summary>
    public static bool SetPrimaryJava(string javaPath) =>
        !string.IsNullOrWhiteSpace(javaPath) &&
        WithStore(store => store.JavaPathSetPrimary(javaPath.Trim()));

    /// <summary>
    /// 全局默认版本隔离设置。true = 新实例默认隔离（mods/config/saves 各实例独立），
    /// false = 默认共享 .minecraft 根目录。未配置时默认 true（启用隔离，更安全、实例互不污染）。
    /// 仅在版本自身的 <c>IsVersionIsolationEnabled</c> 为 null 时生效。
    /// </summary>
    public static bool? DefaultVersionIsolation
    {
        get
        {
            var value = GetValue("defaultVersionIsolation");
            // 缺省启用隔离：避免不同实例的 mods/saves/config 互相污染
            return bool.TryParse(value, out var result) ? result : true;
        }
    }

    /// <summary>保存全局默认版本隔离设置。</summary>
    public static void SaveDefaultVersionIsolation(bool? value)
    {
        if (value.HasValue)
            SetValue("defaultVersionIsolation", value.Value.ToString());
        else
            WithStore(store => store.ConfigItemDelete("defaultVersionIsolation"));
    }

    /// <summary>
    /// 启动前是否校验游戏文件完整性并自动补全缺失文件。默认 true。
    /// </summary>
    public static bool VerifyFilesBeforeLaunch
    {
        get
        {
            var value = GetValue("verifyFilesBeforeLaunch");
            // 未设置时默认开启
            return value is null || bool.TryParse(value, out var result) && result;
        }
    }

    /// <summary>保存启动前文件校验设置。</summary>
    public static void SaveVerifyFilesBeforeLaunch(bool enabled) =>
        SetValue("verifyFilesBeforeLaunch", enabled.ToString());

    /// <summary>保存/更新任意字符串配置项。</summary>
    public static bool SetValue(string key, string value) =>
        !string.IsNullOrWhiteSpace(key) &&
        !string.IsNullOrWhiteSpace(value) &&
        WithStore(store => store.ConfigItemAdd(key, value));

    /// <summary>读取任意字符串配置项；不存在时返回 null。</summary>
    public static string? GetValue(string key)
    {
        return WithStore(store =>
        {
            var value = store.ConfigItemRead(key);
            return string.IsNullOrWhiteSpace(value) ? null : value;
        });
    }

    /// <summary>删除配置项；不存在时无副作用。返回是否删除成功。</summary>
    public static bool ClearValue(string key) =>
        !string.IsNullOrWhiteSpace(key) &&
        WithStore(store => store.ConfigItemDelete(key));

    private static string GetFilePath() =>
        Path.Combine(_storageDirectory, "config.json");

    private static T WithStore<T>(Func<ConfigFileManager, T> action)
    {
        lock (SyncRoot)
        {
            _store ??= new ConfigFileManager(GetFilePath());
            return action(_store);
        }
    }
}

File: internal/config/launcherconfig.go
package config

import (
	"errors"
	"os"
	"path/filepath"
	"strings"
	"sync"
)

// LauncherConfig 启动器配置的统一入口（C# 静态单例类 → Go 包级变量 + 互斥锁）。
// 底层由 ConfigFileManager 负责 JSON 读写。
// 配置文件为存储目录下的 config.json，默认目录与 workspace.json 保持一致
// （%USERPROFILE%\NyaLauncher），可通过 SetStorageDirectory 同步工作区存储目录。

var (
	configSyncRoot   sync.Mutex
	storageDirectory = defaultStorageDirectoryValue()
	sharedStore      *ConfigFileManager
)

// DefaultStorageDirectory 默认存储目录：%USERPROFILE%\NyaLauncher
// （与 workspace.json 默认目录一致，便于用户直接找到并编辑）。
func DefaultStorageDirectory() string { return defaultStorageDirectoryValue() }

func defaultStorageDirectoryValue() string {
	home := UserHome()
	if home == "" {
		return filepath.Join(".", "NyaLauncher")
	}
	return filepath.Join(home, "NyaLauncher")
}

// LegacyDefaultStorageDirectory 旧版默认存储目录：%LOCALAPPDATA%\NyaLauncher。
// 仅用于从旧版本一次性迁移配置到 DefaultStorageDirectory
// （迁移动作由前端的配置存储协调流程发起，见 PORTING_NOTES.md）。
func LegacyDefaultStorageDirectory() string {
	localAppData := os.Getenv("LOCALAPPDATA")
	if localAppData == "" {
		if home := UserHome(); home != "" {
			localAppData = filepath.Join(home, "AppData", "Local")
		}
	}
	if localAppData == "" {
		return filepath.Join(".", "NyaLauncher")
	}
	return filepath.Join(localAppData, "NyaLauncher")
}

// UserHome 用户主目录（%USERPROFILE%，不可用时回落 os.UserHomeDir）。
func UserHome() string {
	if home := os.Getenv("USERPROFILE"); home != "" {
		return home
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return ""
	}
	return home
}

// StorageDirectory config.json 所在目录；默认与 workspace.json 同目录。
func StorageDirectory() string {
	configSyncRoot.Lock()
	defer configSyncRoot.Unlock()
	return storageDirectory
}

// FilePath 配置文件路径：存储目录下的 config.json。
func FilePath() string {
	configSyncRoot.Lock()
	defer configSyncRoot.Unlock()
	return launcherConfigFilePath()
}

// SetStorageDirectory 切换 config.json 的读取目录。文件迁移与冲突处理由前端的
// 配置存储协调流程完成；本方法只重置底层存储，使后续读取立即应用新目录中的配置。
func SetStorageDirectory(storageDir string) error {
	if strings.TrimSpace(storageDir) == "" {
		return errors.New("storageDirectory 不能为空")
	}
	normalized := normalizeDirectory(storageDir)
	configSyncRoot.Lock()
	defer configSyncRoot.Unlock()
	if PathsEqualNormalized(normalized, storageDirectory) {
		return nil
	}
	storageDirectory = normalized
	sharedStore = nil // 下次访问时用新路径重新加载
	return nil
}

// GameDirectory 游戏根目录（.minecraft 或自定义目录）；未配置时返回空串。
func GameDirectory() string {
	return withStoreString(func(store *ConfigFileManager) string {
		value := store.MinecraftPathGet()
		if strings.TrimSpace(value) == "" {
			return ""
		}
		return value
	})
}

// JavaExecutable 首选 Java 可执行文件（java.exe）路径；未配置时返回空串。
func JavaExecutable() string {
	return withStoreString(func(store *ConfigFileManager) string {
		items := store.JavaPathGet()
		if len(items) == 0 {
			return ""
		}
		item := items[0]
		if strings.TrimSpace(item.JavaPath) == "" {
			return ""
		}
		return item.JavaPath
	})
}

// JavaVersion 首选 Java 的版本号（如 21）；未配置时返回空串。
func JavaVersion() string {
	return withStoreString(func(store *ConfigFileManager) string {
		items := store.JavaPathGet()
		if len(items) == 0 {
			return ""
		}
		item := items[0]
		if strings.TrimSpace(item.JavaVersion) == "" {
			return ""
		}
		return item.JavaVersion
	})
}

// SaveGameDirectory 保存游戏目录。
func SaveGameDirectory(path string) bool {
	if strings.TrimSpace(path) == "" {
		return false
	}
	return withStoreBool(func(store *ConfigFileManager) bool {
		return store.MinecraftPathSet(strings.TrimSpace(path))
	})
}

// ClearGameDirectory 清除已保存的游戏目录（恢复自动检测）。
func ClearGameDirectory() {
	withStoreBool(func(store *ConfigFileManager) bool {
		store.ConfigItemDelete("minecraftPath")
		return true
	})
}

// SaveJava 保存首选 Java（java.exe 路径 + 版本）。采用「先清空再写入」策略，
// 保证 config.json 中的 javaPath 始终只有一条首选配置。
func SaveJava(javaPath, javaVersion string) bool {
	if strings.TrimSpace(javaPath) == "" {
		return false
	}
	if strings.TrimSpace(javaVersion) == "" {
		javaVersion = "unknown"
	}
	return withStoreBool(func(store *ConfigFileManager) bool {
		return store.JavaPathSet(strings.TrimSpace(javaPath), strings.TrimSpace(javaVersion))
	})
}

// ClearJava 删除已保存的 Java 配置（恢复自动检测）。
func ClearJava() {
	withStoreBool(func(store *ConfigFileManager) bool {
		store.ConfigItemDelete("javaPath")
		return true
	})
}

// GetJavaPaths 已保存的全部 Java 路径（列表首位为默认 Java）。
func GetJavaPaths() []JavaPathItem {
	configSyncRoot.Lock()
	defer configSyncRoot.Unlock()
	store := ensureStore()
	if store == nil {
		return []JavaPathItem{}
	}
	return store.JavaPathGet()
}

// AddJava 添加一条 Java 路径；路径已存在时更新其版本。返回是否成功。
func AddJava(javaPath, javaVersion string) bool {
	if strings.TrimSpace(javaPath) == "" {
		return false
	}
	if strings.TrimSpace(javaVersion) == "" {
		javaVersion = "unknown"
	}
	return withStoreBool(func(store *ConfigFileManager) bool {
		return store.JavaPathAdd(strings.TrimSpace(javaPath), strings.TrimSpace(javaVersion))
	})
}

// RemoveJava 移除一条 Java 路径。返回是否移除成功。
func RemoveJava(javaPath string) bool {
	if strings.TrimSpace(javaPath) == "" {
		return false
	}
	return withStoreBool(func(store *ConfigFileManager) bool {
		return store.JavaPathRemove(strings.TrimSpace(javaPath))
	})
}

// SetPrimaryJava 把指定路径设为默认 Java（列表首位）。返回是否成功。
func SetPrimaryJava(javaPath string) bool {
	if strings.TrimSpace(javaPath) == "" {
		return false
	}
	return withStoreBool(func(store *ConfigFileManager) bool {
		return store.JavaPathSetPrimary(strings.TrimSpace(javaPath))
	})
}

// DefaultVersionIsolation 全局默认版本隔离设置。true = 新实例默认隔离
// （mods/config/saves 各实例独立），false = 默认共享 .minecraft 根目录。
// 未配置时默认 true（启用隔离，更安全、实例互不污染）。
// 仅在版本自身的 IsVersionIsolationEnabled 为 null 时生效。
func DefaultVersionIsolation() bool {
	value := GetValue("defaultVersionIsolation")
	// 缺省启用隔离：避免不同实例的 mods/saves/config 互相污染
	result, err := parseBool(value)
	if err != nil {
		return true
	}
	return result
}

// SaveDefaultVersionIsolation 保存全局默认版本隔离设置；value 为 nil 时删除配置项。
func SaveDefaultVersionIsolation(value *bool) {
	if value != nil {
		setValue("defaultVersionIsolation", formatBool(*value))
		return
	}
	withStoreBool(func(store *ConfigFileManager) bool {
		store.ConfigItemDelete("defaultVersionIsolation")
		return true
	})
}

// VerifyFilesBeforeLaunch 启动前是否校验游戏文件完整性并自动补全缺失文件。默认 true。
func VerifyFilesBeforeLaunch() bool {
	value := GetValue("verifyFilesBeforeLaunch")
	// 未设置时默认开启
	if value == "" {
		return true
	}
	result, err := parseBool(value)
	if err != nil {
		return false
	}
	return result
}

// SaveVerifyFilesBeforeLaunch 保存启动前文件校验设置。
func SaveVerifyFilesBeforeLaunch(enabled bool) {
	setValue("verifyFilesBeforeLaunch", formatBool(enabled))
}

// SetValue 保存/更新任意字符串配置项。
func SetValue(key, value string) bool {
	if strings.TrimSpace(key) == "" || strings.TrimSpace(value) == "" {
		return false
	}
	return withStoreBool(func(store *ConfigFileManager) bool {
		return store.ConfigItemAdd(key, value)
	})
}

// setValue 包内小写入口：跳过外部重复校验（仅供包内已校验过的调用使用）。
func setValue(key, value string) bool {
	return SetValue(key, value)
}

// GetValue 读取任意字符串配置项；不存在（或为空白）时返回空串。
func GetValue(key string) string {
	return withStoreString(func(store *ConfigFileManager) string {
		value := store.ConfigItemRead(key)
		if strings.TrimSpace(value) == "" {
			return ""
		}
		return value
	})
}

// ClearValue 删除配置项；不存在时无副作用。返回是否删除成功。
func ClearValue(key string) bool {
	if strings.TrimSpace(key) == "" {
		return false
	}
	return withStoreBool(func(store *ConfigFileManager) bool {
		return store.ConfigItemDelete(key)
	})
}

// UpdateInTransaction 在单个原子写入中应用多项配置修改：任一环节失败时整体回滚并放弃落盘。
// 回调直接操作 JSON 文档，不要在其中调用本类的 SetValue 等方法（会破坏事务性）。
func UpdateInTransaction(mutation func(map[string]any) bool) bool {
	return withStoreBool(func(store *ConfigFileManager) bool {
		return store.UpdateInTransaction(mutation)
	})
}

func launcherConfigFilePath() string {
	return filepath.Join(storageDirectory, "config.json")
}

// ensureStore 取得（或按需创建）底层 ConfigFileManager，需持 configSyncRoot 调用。
func ensureStore() *ConfigFileManager {
	if sharedStore == nil {
		store, err := NewConfigFileManager(launcherConfigFilePath())
		if err != nil {
			logsWriteError("初始化配置存储失败: " + err.Error())
			return nil
		}
		sharedStore = store
	}
	return sharedStore
}

func withStoreBool(action func(*ConfigFileManager) bool) bool {
	configSyncRoot.Lock()
	defer configSyncRoot.Unlock()
	store := ensureStore()
	if store == nil {
		return false
	}
	return action(store)
}

func withStoreString(action func(*ConfigFileManager) string) string {
	configSyncRoot.Lock()
	defer configSyncRoot.Unlock()
	store := ensureStore()
	if store == nil {
		return ""
	}
	return action(store)
}

// normalizeDirectory 规范化目录：展开完整路径并去掉尾部目录分隔符。
func normalizeDirectory(path string) string {
	trimmed := strings.TrimSpace(path)
	abs, err := filepath.Abs(trimmed)
	if err != nil {
		return trimTrailingSeparator(trimmed)
	}
	return trimTrailingSeparator(filepath.Clean(abs))
}

// PathsEqualNormalized 已规范化路径的比较：Windows 忽略大小写。
func PathsEqualNormalized(left, right string) bool {
	return pathsEqualFold(left, right)
}
